File: packages/sdk-environment/src/sdk_environment/_telemetry.py
from __future__ import annotations

from opentelemetry import trace
from opentelemetry.trace import Span, Status, StatusCode

from ._types import EnvironmentFailure, StructuredEvent
from ._version import ENVIRONMENT_SDK_VERSION

_TRACER_NAME = "meridian.sdk-environment"


def get_tracer() -> trace.Tracer:
    return trace.get_tracer(_TRACER_NAME, ENVIRONMENT_SDK_VERSION)


def record_invocation_event(span: Span, event: StructuredEvent) -> None:
    """
    Attaches a structured "environment.invocation" event to the active span.
    Called once per runtime operation regardless of success or failure.
    """
    attrs: dict[str, str | int | float | bool] = {}
    for k, v in vars(event).items():
        if isinstance(v, (str, int, float, bool)):
            attrs[k] = v
    span.add_event("environment.invocation", attrs)


def record_environment_failure(span: Span, failure: EnvironmentFailure) -> None:
    """
    Records a failure on the span: sets status to ERROR, adds an
    "environment.error" event, and records the underlying exception if present.
    """
    span.set_status(Status(StatusCode.ERROR, failure.message))
    span.add_event(
        "environment.error",
        {
            "environment.id": failure.environment_id,
            "environment.kind": failure.environment_kind,
            "session.id": failure.session_id,
            "error.code": failure.code,
            "error.message": failure.message,
        },
    )
    if failure.cause is not None:
        span.record_exception(failure.cause)

File: packages/sdk-widget/src/telemetry.ts
import { SpanStatusCode, trace } from "@opentelemetry/api";
import type { Attributes, Span } from "@opentelemetry/api";
import type { StructuredEvent, WidgetError } from "./types.js";
import { WIDGET_SDK_VERSION } from "./version.js";

const TRACER_NAME = "meridian.sdk-widget";

export function getTracer() {
  return trace.getTracer(TRACER_NAME, WIDGET_SDK_VERSION);
}

/**
 * Attaches a structured "widget.invocation" event to the active span.
 * Called once per pipeline invocation regardless of success or failure.
 */
export function recordInvocationEvent(span: Span, event: StructuredEvent): void {
  const attrs: Attributes = {};
  for (const [k, v] of Object.entries(event)) {
    if (typeof v === "string" || typeof v === "number" || typeof v === "boolean") {
      attrs[k] = v;
    }
  }
  span.addEvent("widget.invocation", attrs);
}

/**
 * Records a widget failure on the span: sets status to ERROR,
 * adds a "widget.error" event, and records the underlying exception if present.
 */
export function recordWidgetFailure(span: Span, error: WidgetError): void {
  span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
  span.addEvent("widget.error", {
    "widget.id": error.widget_id,
    "widget.kind": error.widget_kind,
    "session.id": error.session_id,
    "error.code": error.code,
    "error.message": error.message,
  });
  if (error.cause instanceof Error) {
    span.recordException(error.cause);
  }
}
